src/synapse_api.py

```diff
@@ -0,0 +1,178 @@
+
+SYNAPSE_URL = "http://localhost:8008"
+URL_REGISTER = f"{SYNAPSE_URL}/_synapse/admin/v1/register"
+
+
+class RegisterUserError(Exception):
+    """Exception raised when registering user via API fails.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the RegisterUserError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class NetworkError(Exception):
+    """Exception raised when requesting API fails due network issues.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the NetworkError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class SynapseAPI:
+    """The Synapse API handler."""
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the handler.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        self._charm = charm
+
+    def _get_nonce(self) -> str:
+        """Get nonce.
+
+        Returns:
+            The nonce returned by Synapse API.
+
+        Raises:
+            NetworkError: if there was an error fetching the nonce.
+        """
+        try:
+            res = requests.get(URL_REGISTER, timeout=5)
+            res.raise_for_status()
+            return res.json()["nonce"]
+        except (
+            requests.exceptions.ConnectionError,
+            requests.exceptions.Timeout,
+            requests.exceptions.HTTPError,
+        ) as exc:
+            logger.error("Failed to request %s : %s", URL_REGISTER, exc)
+            raise NetworkError(f"Failed to request {URL_REGISTER}.") from exc
+
+    def _generate_mac(
+        self,
+        shared_secret: str,
+        nonce: str,
+        user: str,
+        password: str,
+        admin: bool = False,
+        user_type: typing.Optional[str] = None,
+    ) -> str:
+        """Generate mac to register user.
+
+        Args:
+            shared_secret: registration_shared_secret from configuration file.
+            nonce: nonce generated.
+            user: username.
+            password: password.
+            admin: if is admin. Default False.
+            user_type: user type. Defaults to None.
+
+        Returns:
+            _type_: _description_
+        """
+        mac = hmac.new(key=shared_secret.encode("utf8"), digestmod=hashlib.sha1)
+        mac.update(nonce.encode("utf8"))
+        mac.update(b"\x00")
+        mac.update(user.encode("utf8"))
+        mac.update(b"\x00")
+        mac.update(password.encode("utf8"))
+        mac.update(b"\x00")
+        mac.update(b"admin" if admin else b"notadmin")
+        if user_type:
+            mac.update(b"\x00")
+            mac.update(user_type.encode("utf8"))
+
+        return mac.hexdigest()
```

Comment:
Is there a particular reason for the order and adding `b"\x00"` in between?

Reply:
This is what the [documentation](https://matrix-org.github.io/synapse/v1.41/admin_api/register_api.html) suggests to do. The same happens in the [script](https://github.com/matrix-org/synapse/blob/develop/synapse/_scripts/register_new_matrix_user.py#L68).

Reply:
ok, do you think it should be documented with links here? I'm guessing a future dev would be as confused as I am 🙂

Reply:
Sure!! Nice suggestion :)